### ISSUE TITLE
ESP8266: option to use NMI for PWM

### ISSUE THREAD

**Issue by Icaltary** [micropython/micropython#6555]
Hi,
I detected some instabilities in the PWM output of the ESP8266 when the controller is handling WiFi traffic. Of course, that's no surprise, since the PWM is done in software. However, I figured out, that the issues can be mostly mitigated by using the NMI interrupt for the PWM. I guess, that's not something we want to be default, but as an option it would be nice.

The issue can be replicated by setting up a PWM output with an LED at a duty cycle of 50%. Then a simple script is needed, that echoes data on a TCP socket. When lots of data is thrown at the ESP, the LED will begin to flicker visibly.
The issue easiest to reproduce with heavy network traffic, but it will also occur more sporadically in lighter traffic environments.

Further, I'd suggest to replace the gpio_output_set calls in the interrupt handler with direct writes to the W1TC / W1TC registers to reduce time wasted in the interrupt.

**Comment by xuancong84** [comment]
Yup, I also found ESP8266 Micropython PWM flickering issue. I use PWM on AMS1117 to control brightness of LED beads. And the flickering exists even when Wifi is inactive, though much milder than when Wifi is active. On the other hand, my C++ version smart-light controller program (on the same ESP8266 chip) does not have this issue even when Wifi is active. So this is definitely a defect in MicroPython.
